### PR TITLE
kernel: Add NCS boot banner

### PR DESCRIPTION
**Pull request by nordicjm**
Adds an enhanced boot banner which can be configured to show application name and version, NCS version, and underlying zephyr version.

### CODE REVIEW

**Comment by nordicjm** [comment]
@gchwier This looks to need an update to the `test-fw-nrfconnect-boot` tests, though would hold off until people have approved the PR

File: kernel/banner.c
Comment (tejlmand) [inline kernel/banner.c:1]: This file contains same function names as its upstream counter part.

But the upstream `boot_banner.c` function is not upstream.
Both versions of `banner.c` are compiled and linked, which means that the linker picks the first function which satisfies linking. This happens to work because nrf banner.c is added to zephyr lib, which is inside the whole-archive flags, and the other is outside (libkernel.a).

This is a risky approach, which I don't like.
I think upstream `banner.c` should instead be made such a way that a downstream project can easily replace it.

From build:
```

[37/145] ccache /opt/zephyr-sdk-0.16.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc ... -MF zephyr/CMakeFiles/zephyr.dir/projects/github/ncs/nrf/kernel/banner.c.obj.d -o zephyr/CMakeFiles/zephyr.dir/projects/github/ncs/nrf/kernel/banner.c.obj -c /projects/github/ncs/nrf/kernel/banner.c
...
[105/145] ccache /opt/zephyr-sdk-0.16.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc ... -MF zephyr/kernel/CMakeFiles/kernel.dir/banner.c.obj.d -o zephyr/kernel/CMakeFiles/kernel.dir/banner.c.obj -c /projects/github/ncs/zephyr/kernel/banner.c
...
[144/145] : && ccache /opt/zephyr-sdk-0.16.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc  ... -o zephyr/zephyr.elf  ...  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  ... -Wl,--no-whole-archive  zephyr/kernel/libkernel.a ...
```

File: cmake/version.cmake
Comment (tejlmand) [inline cmake/version.cmake]: A lot of duplication here.

```suggestion
set(types ncs zephyr)
set(types_upper NCS ZEPHYR)
set(path_vars NRF_DIR ZEPHYR_BASE)

if(${APP_VERSION_NUMBER})
  set(types ${types} app)
  set(types_upper ${types_upper} APP)
  set(path_vars ${path_vars} CMAKE_SOURCE_DIR)
endif()

foreach(type IN ZIP_LISTS types types_upper path_vars)
  add_custom_command(
    OUTPUT ${PROJECT_BINARY_DIR}/include/generated/${type_0}_commit.h
    COMMAND ${CMAKE_COMMAND} -DZEPHYR_BASE=${ZEPHYR_BASE} -DNRF_DIR=${NRF_DIR}
      -DOUT_FILE=${PROJECT_BINARY_DIR}/include/generated/${type_0}_commit.h
      -DCOMMIT_TYPE=${type_1}
      -DCOMMIT_PATH=${${type_2}}
      -P ${ZEPHYR_NRF_MODULE_DIR}/cmake/gen_commit_h.cmake
      DEPENDS ${${type_2}}/VERSION ${git_dependency}
  )
  add_custom_target(ncs_commit_h DEPENDS ${PROJECT_BINARY_DIR}/include/generated/ncs_commit.h)
  add_dependencies(version_h ${type_0}_commit_h)
endforeach()
```

Note, it's possible to use `foreach(type TUPPER path_var IN ZIP_LISTS types types_upper path_vars)` and then refer to those as `type` (was: type_0), `TUPPER` (was: type_1), `path_var` (was: type_2) in above code.

File: cmake/gen_commit_h.cmake
Comment (tejlmand) [inline cmake/gen_commit_h.cmake:31]: this will not work as expected in all cases.
Please keep out of this PR and open separately for further discussions.

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: cmake/gen_commit_h.cmake
Comment (nordicjm) [reply cmake/gen_commit_h.cmake:31]: in what cases will it not work?

Comment (nordicjm) [reply cmake/gen_commit_h.cmake]: Has been removed

File: kernel/CMakeLists.txt
Comment (nordicjm) [reply kernel/CMakeLists.txt]: Fixed

**Review by LuDuda — APPROVED** [review]
This is great! :+1:

File: cmake/version_app.cmake
Comment (tejlmand) [inline cmake/version_app.cmake:1]: why do we need a dedicated `version_app.cmake`.

Also, if we do need it, then it should have copyright and license header.

File: kernel/banner.c
Comment (tejlmand) [inline kernel/banner.c:39]: instead of this, wouldn't it be better to indicate that there was no git repo found ?

I believe something like this:
```
*** Booting my_app v1.0.0 - unknown commit
```

will catch attention earlier that something might be wrong, compared to:
```
*** Booting my_app v1.0.0
```

(None, Null, empty, etc are also options)

Comment (nordicjm) [reply kernel/banner.c:39]: It is omitted because it might not be in a git repo

File: cmake/version_app.cmake
Comment (nordicjm) [reply cmake/version_app.cmake:1]: Because the user application might not reside in a git repo, however zephyr and ncs do

File: kernel/banner.c
Comment (tejlmand) [reply kernel/banner.c:39]: Exactly, as I wrote:
> wouldn't it be better to indicate that there was no git repo found ?

So I understand why it's omitted, what i'm asking is to indicate that no git repo was found, exactly for this reason:
> will catch attention earlier that something might be wrong, 

Feel free to decide how to indicate that the app is not in a git repo, but I think it should be indicated.

File: cmake/version_app.cmake
Comment (tejlmand) [reply cmake/version_app.cmake:7]: > Because the user application might not reside in a git repo, however zephyr and ncs do

I don't see why the question of application being in a git repo or not should decide whether the version files cannot be unified ?

If this is because you think the `git` handling doesn't belong in `version.cmake` then I prefer that upstream is updated to support a common `git` CMake function, which can then be reused downstream, and only the need for a single `version.cmake` file.

File: kernel/banner.c
Comment (nordicjm) [reply kernel/banner.c:39]: Added

File: cmake/version_app.cmake
Comment (nordicjm) [reply cmake/version_app.cmake:7]: It is need to know if a commit file is generated and what message to show

**Comment by nordicjm** [comment]
@gchwier Would you be able to prepare updates to the test-fw-nrfconnect-boot tests for this change?

**Comment by gchwier** [comment]
> @gchwier Would you be able to prepare updates to the test-fw-nrfconnect-boot tests for this change?

I've updated it two weeks ago. Now regression test on test-fw-nrfconnect-boot are passed (previous fail was because of some problems in lab)

**Review by tejlmand — APPROVED** [review]

**Pull request merged**
